### ISSUE TITLE
CTC bug with multi-gpu training

### ISSUE THREAD

**Issue by NirHeaven** [pytorch/pytorch#12201]
I am using pytorch 0.5 compiled from github. but when I training a s2s network on multi-gpu with some AMAZING batchsize(2 gpu with batchsize greater than 128) using CTC loss, the loss will continue to rise!!!But when I choose a smaller batchsize, everything works well!

**Comment by t-vi** [comment]
I must admit that I don't have much of an idea what happens here, and I cannot test multi-gpu at the moment. There is a switch between GPU backward kernels that depends on the size. Are you sure it only happens with multiple GPU?

When you say 0.5, do you mean current master (which should report 1.0a)? I fixed some bugs a month ago or so.

**Comment by jinserk** [comment]
Hi @t-vi,

I have a little different question for the newly added `CTCLoss`. I think the 'elementwise_mean' option is the same to be averaging along the batch and the sequence length, but `nn.CTCLoss` produce almost 10x higher loss than @SeanNaren 's warp-ctc pytorch binding. In my understanding, the interface is almost the same, so I just replace the `warpctc_pytorch.CTCLoss` to `nn.CTCLoss`, but the result was so. Is it normal? The loss reduction in training is obviously not so smooth as warp-ctc did. I had tried Adam instead of SGD what I did use as the optimizer, which makes the convergence better, but still the slope is very low compared to the previous one.

**Comment by t-vi** [comment]
I think this might be the averaging - PyTorch uses "nats per target character" akin to the bpc metric commonly seen for character rnns, Sean might divide by the input length. Heuristically, you should scale down the learning rate to compensate for the change in error scale unless your optimizer normalizes out the scale.
It might be more appropriate to discuss on the forum, though.

**Comment by jinserk** [comment]
Thanks @t-vi ! I agree with you that the forum is the better place to discuss, but this CTCLoss is very recent features that few people have experienced, including me. I didn't know that higher loss was an intended one or a bug. So now understand that the higher loss is originated from the fact that it was normalized by the target lengths, not by the input lengths. Actually I had tried lower initial learning rate but the training was still unstable. I'll do more trials with the parameters. Thanks!

**Comment by jinserk** [comment]
@t-vi , could you let me ask one more question? When I use the native CTCLoss in pytorch, most of the convergences were the sequence of blanks -- but they still have a very low loss, so they were not trained well. I didn't include blanks when I made the target label sequences. Did I do something wrong?

**Comment by t-vi** [comment]
@jinserk Look, I'd be more than happy to look into this, but the [forum](https://discuss.pytorch.org/) would be a much better place than a random bug report and you'd probably have much more success with "I have the following inputs and targets, warp_ctc does X, PyTorch CTC does Y, for me X is much better".

**Comment by jinserk** [comment]
@t-vi sorry if I make you annoying. I opened a thread of this issue in the forum [here](https://discuss.pytorch.org/t/ctcloss-performance-of-pytorch-1-0-0/27524).

**Comment by gchanan** [comment]
closing because this is old and no reproduction provided.
